fix(docs): documentation about resource name generation (#1225)
cdk8s-team/cdk8s: docs/basics/api-object.md

`--class-prefix` flag, or by passing `--no-class-prefix`. (Note: if no prefix is
used for k8s resources, this may lead to conflicts - see
https://github.com/cdk8s-team/cdk8s/issues/140).

### Resource Names

> Every object has a name that is autogenerated when the cdk8s app is synthesized.

Reviewer: Instead of `object` this can be `resource`.

The name is the passed in string you assign to an object, prefixed by the name of the
chart it is contained in, and suffixed by the hash of the construct path.

For example, the following code:

```ts
import { App, Chart } from 'cdk8s';
import * as kplus from 'cdk8s-plus-25';

const app = new App();
const chart = new MyChart(app, 'my-chart');

new kplus.Deployment(chart, 'deployment', {
  containers: [{ image: 'nginx' }]
});

app.synth();
```

Generates a `Deployment` with the name: `my-chart-deployment-c8c354dd`.

Hashes are used to ensure that all resources have unique identifiers. However, they are
not required and it may sometimes be undesirable to include the hash suffixes.

To remove hashes from resource names, you can change the above code by setting the
`ChartProps.disableResourceNameHashes` value to `true` like so:

```ts
import { App, Chart } from 'cdk8s';
import * as kplus from 'cdk8s-plus-25';

const app = new App();
const chart = new MyChart(app, 'my-chart', {
    disableResourceNameHashes: true,
});

new kplus.Deployment(chart, 'deployment', {
  containers: [{ image: 'nginx' }]
});

app.synth();
```

> Which would similarly generate a `Deployment` now with the name: `my-chart-deployment`.

Reviewer: Have we mentioned anywhere what issues can non unique identifier cause for the customer?

Author: Added a line about how unique names are required in K8s - let me know if this is sufficient

`--class-prefix` flag, or by passing `--no-class-prefix`. (Note: if no prefix is
used for k8s resources, this may lead to conflicts - see
https://github.com/cdk8s-team/cdk8s/issues/140).

### Resource Names

> Every resource has a name that is autogenerated when the cdk8s app is synthesized.
> The name is the passed in string you assign to an object, prefixed by the name of the
> chart it is contained in, and suffixed by the hash of the construct path.

Reviewer 2: ```suggestion
By default, cdk8s will autogenerate a name for every resource when it is instantiated.
The name is created from the path of the construct, and suffixed with a hash of that path.
```

For example, the following code:

```ts
import { App, Chart } from 'cdk8s';
import * as kplus from 'cdk8s-plus-25';

const app = new App();
const chart = new MyChart(app, 'my-chart');

new kplus.Deployment(chart, 'deployment', {
  containers: [{ image: 'nginx' }]
});

app.synth();
```

> Generates a `Deployment` with the name: `my-chart-deployment-c8c354dd`.

Reviewer 2: ```suggestion
Generates a `Deployment` with the name: `my-chart-deployment-c8c354dd`. Where `my-chart` is the id of the `Chart`, `deployment` is the id of the `Deployment`, and `c8c354dd` is the hash suffix.
```

> In Kubernetes, only one resource of a given kind can have a given name at a time.
> Hashes are used to ensure that all resources have unique identifiers. However, they are
> not required and it may sometimes be undesirable to include the hash suffixes.

Reviewer 2: ```suggestion
You can programmatically access the name of any resource in a cdk8s app by accessing the `.name` property of a resource. For example:

```ts
const deployment = new kplus.Deployment(chart, 'deployment', {
  containers: [{ image: 'nginx' }]
});
const deploymentName = deployment.name;
```
```

> To remove hashes from resource names, you can change the above code by setting the
> `ChartProps.disableResourceNameHashes` value to `true` like so:

Reviewer 2: ```suggestion
To remove hashes from resource names, you can set the `disableResourceNameHashes` property to `true`:
```

```ts

> import { App, Chart } from 'cdk8s';
> import * as kplus from 'cdk8s-plus-25';
> 
> const app = new App();
> const chart = new MyChart(app, 'my-chart', {
>     disableResourceNameHashes: true,
> });
> 
> new kplus.Deployment(chart, 'deployment', {
>   containers: [{ image: 'nginx' }]
> });
> 
> app.synth();

Reviewer 2: ```suggestion
...
const chart = new MyChart(app, 'my-chart', {
    disableResourceNameHashes: true,
});
...
```

```

> Which would similarly generate a `Deployment` now with the name: `my-chart-deployment`.

Reviewer 2: ```suggestion
Which would similarly generate a `Deployment` with the name: `my-chart-deployment`. Note that disabling hashes introduces the possibility of name clashes. For example, the following code:

```ts
// think of this as a custom construct
const c = new Construct(this, 'Construct');

new kplus.Deployment(c, 'Deployment', {
  containers: [{ image: 'image' }]
});

new kplus.Deployment(this, 'Construct-Deployment', {
  containers: [{ image: 'image' }]
})
```

Will produce two deployments with the same name: `chart-construct-deployment`. This happens because cdk8s uses a hyphen (`-`) as the delimiter between elements in the path, so if the id of your construct also contains a hyphen (`-`), a collision can occur.

In addition, you can explicitly set a name for every resource by using the `metadata.name` property:

```ts
new kplus.Deployment(c, 'Deployment', {
  metadata: {
    name: 'my-deployment'
  },
  ...
});
```

```
